### ISSUE TITLE
bug: prisms with symmetry in Z direction

### ISSUE THREAD

**Issue by oskooi** [stevengj/meep#440]
There is a bug in the `Prism` object when used with symmetry in the `Z` direction. It works fine with symmetry in the `X` and `Y` directions. This is demonstrated in the simulation script below which involves a hexagonal prism (a convex shape is intentionally chosen so as not to be affected by a separate bug for non-convex shapes described in #400). A point dipole source is placed at the center of the computational cell and the field at a random position is output at every 5 time units. Results from three different runs are compared: (1) no symmetry (`use_symmetries = False`), (2) mirror symmetry in `X` and `Y` (`use_symmetries = True` and `use_z_symmetry = False`), and (3) mirror symmetry in `X`, `Y`, and `Z` (`use_symmetries = True` and `use_z_symmetry = True`). The first two runs produce nearly identical output but the third is significantly different.

**Simulation Script**
```py
import meep as mp
import math

cell_size = mp.Vector3(5,5,5)

Si = mp.Medium(index=3.5)

vertices = [mp.Vector3(-1,0),
            mp.Vector3(-0.5,math.sqrt(3)/2),
            mp.Vector3(+0.5,math.sqrt(3)/2),
            mp.Vector3(1,0),
            mp.Vector3(+0.5,-math.sqrt(3)/2),
            mp.Vector3(-0.5,-math.sqrt(3)/2)]

geometry = [mp.Prism(vertices, height=1.0, material=Si)]

boundary_layers = [mp.PML(thickness=1.0)]

fcen = 1.0
sources = [mp.Source(mp.GaussianSource(fcen, fwidth=0.2*fcen), component=mp.Ez, center=mp.Vector3())]

use_symmetries = False

use_z_symmetry = True

if use_symmetries:
    if use_z_symmetry:
        symmetries = [mp.Mirror(direction=mp.X,phase=+1),
                      mp.Mirror(direction=mp.Y,phase=+1),
                      mp.Mirror(direction=mp.Z,phase=-1)]
    else:
        symmetries = [mp.Mirror(direction=mp.X,phase=+1),
                      mp.Mirror(direction=mp.Y,phase=+1)]
else:
    symmetries = []

sim = mp.Simulation(resolution=30,
                    geometry=geometry,
                    cell_size=cell_size,
                    boundary_layers=boundary_layers,
                    sources=sources,
                    symmetries=symmetries)

def print_stuff(sim):
    p = sim.get_field_point(mp.Ex, mp.Vector3(-0.21,0.37,0.18))
    print("ex:, {}, {}".format(sim.meep_time(), p.real))

sim.run(mp.at_beginning(mp.output_epsilon), mp.at_every(5,print_stuff), until_after_sources=0)
```

**Output (no symmetry)**
```
Using MPI version 3.0, 1 processes
-----------
Initializing structure...
Working in 3D dimensions.
Computational cell is 5 x 5 x 5 with resolution 30
     prism, center = (0,0,0.5)
          height 1, axis (0,0,1), 6 vertices:
          {-1.000000e+00,5.551115e-17,0.000000e+00}
          {-5.000000e-01,8.660254e-01,0.000000e+00}
          {5.000000e-01,8.660254e-01,0.000000e+00}
          {1.000000e+00,-5.551115e-17,0.000000e+00}
          {5.000000e-01,-8.660254e-01,0.000000e+00}
          {-5.000000e-01,-8.660254e-01,0.000000e+00}
          dielectric constant epsilon diagonal = (12.25,12.25,12.25)
time for set_epsilon = 5.53754 s
-----------
Meep progress: 0.05/50.0 = 0.1% done in 4.0s, 4044.3s to go
on time step 3 (time=0.05), 1.34947 s/step
Meep progress: 1.0333333333333332/50.0 = 2.1% done in 8.1s, 383.4s to go
on time step 62 (time=1.03333), 0.0685181 s/step
Meep progress: 2.0/50.0 = 4.0% done in 12.1s, 291.0s to go
on time step 120 (time=2), 0.0695824 s/step
Meep progress: 2.933333333333333/50.0 = 5.9% done in 16.1s, 258.9s to go
on time step 176 (time=2.93333), 0.0715693 s/step
Meep progress: 3.8833333333333333/50.0 = 7.8% done in 20.2s, 239.5s to go
on time step 233 (time=3.88333), 0.070824 s/step
Meep progress: 4.816666666666666/50.0 = 9.6% done in 24.2s, 226.8s to go
on time step 289 (time=4.81667), 0.0715955 s/step
ex:, 5.0, -2.7087240517591588e-05
Meep progress: 5.75/50.0 = 11.5% done in 28.2s, 217.0s to go
on time step 345 (time=5.75), 0.0718329 s/step
Meep progress: 6.683333333333334/50.0 = 13.4% done in 32.2s, 208.9s to go
on time step 401 (time=6.68333), 0.0718922 s/step
Meep progress: 7.633333333333333/50.0 = 15.3% done in 36.3s, 201.2s to go
on time step 458 (time=7.63333), 0.0706855 s/step
Meep progress: 8.583333333333334/50.0 = 17.2% done in 40.3s, 194.6s to go
on time step 515 (time=8.58333), 0.0713371 s/step
Meep progress: 9.516666666666666/50.0 = 19.0% done in 44.3s, 188.6s to go
on time step 571 (time=9.51667), 0.0715376 s/step
ex:, 10.0, -0.0011262756541097275
Meep progress: 10.45/50.0 = 20.9% done in 48.4s, 183.1s to go
on time step 627 (time=10.45), 0.0724217 s/step
Meep progress: 11.4/50.0 = 22.8% done in 52.4s, 177.6s to go
on time step 684 (time=11.4), 0.0711105 s/step
Meep progress: 12.35/50.0 = 24.7% done in 56.5s, 172.2s to go
on time step 741 (time=12.35), 0.0709922 s/step
Meep progress: 13.299999999999999/50.0 = 26.6% done in 60.6s, 167.1s to go
on time step 798 (time=13.3), 0.0713269 s/step
Meep progress: 14.25/50.0 = 28.5% done in 64.6s, 162.1s to go
on time step 855 (time=14.25), 0.0711571 s/step
ex:, 15.0, -0.017212050979577814
Meep progress: 15.2/50.0 = 30.4% done in 68.6s, 157.2s to go
on time step 912 (time=15.2), 0.0708952 s/step
Meep progress: 16.133333333333333/50.0 = 32.3% done in 72.7s, 152.5s to go
on time step 968 (time=16.1333), 0.071633 s/step
Meep progress: 17.066666666666666/50.0 = 34.1% done in 76.7s, 148.0s to go
on time step 1024 (time=17.0667), 0.0717371 s/step
Meep progress: 18.016666666666666/50.0 = 36.0% done in 80.7s, 143.3s to go
on time step 1081 (time=18.0167), 0.0707885 s/step
Meep progress: 18.966666666666665/50.0 = 37.9% done in 84.7s, 138.7s to go
on time step 1138 (time=18.9667), 0.0707981 s/step
Meep progress: 19.916666666666668/50.0 = 39.8% done in 88.8s, 134.1s to go
on time step 1195 (time=19.9167), 0.0711264 s/step
ex:, 20.0, -0.09465803368773307
Meep progress: 20.85/50.0 = 41.7% done in 92.8s, 129.8s to go
on time step 1251 (time=20.85), 0.0719452 s/step
Meep progress: 21.78333333333333/50.0 = 43.6% done in 96.9s, 125.5s to go
on time step 1307 (time=21.7833), 0.0718986 s/step
Meep progress: 22.733333333333334/50.0 = 45.5% done in 100.9s, 121.0s to go
on time step 1364 (time=22.7333), 0.0709691 s/step
Meep progress: 23.683333333333334/50.0 = 47.4% done in 104.9s, 116.6s to go
on time step 1421 (time=23.6833), 0.0708871 s/step
Meep progress: 24.633333333333333/50.0 = 49.3% done in 109.0s, 112.2s to go
on time step 1478 (time=24.6333), 0.0706549 s/step
ex:, 25.0, -0.17097397944148557
Meep progress: 25.583333333333332/50.0 = 51.2% done in 113.0s, 107.9s to go
on time step 1535 (time=25.5833), 0.070804 s/step
Meep progress: 26.53333333333333/50.0 = 53.1% done in 117.0s, 103.5s to go
on time step 1592 (time=26.5333), 0.070725 s/step
Meep progress: 27.466666666666665/50.0 = 54.9% done in 121.1s, 99.3s to go
on time step 1648 (time=27.4667), 0.0717912 s/step
Meep progress: 28.416666666666668/50.0 = 56.8% done in 125.1s, 95.0s to go
on time step 1705 (time=28.4167), 0.0712659 s/step
Meep progress: 29.366666666666667/50.0 = 58.7% done in 129.2s, 90.7s to go
on time step 1762 (time=29.3667), 0.070721 s/step
ex:, 30.0, -0.03653527198458597
Meep progress: 30.316666666666666/50.0 = 60.6% done in 133.2s, 86.5s to go
on time step 1819 (time=30.3167), 0.0706682 s/step
Meep progress: 31.25/50.0 = 62.5% done in 137.2s, 82.3s to go
on time step 1875 (time=31.25), 0.0716984 s/step
Meep progress: 32.18333333333333/50.0 = 64.4% done in 141.2s, 78.2s to go
on time step 1931 (time=32.1833), 0.071506 s/step
Meep progress: 33.13333333333333/50.0 = 66.3% done in 145.3s, 73.9s to go
on time step 1988 (time=33.1333), 0.0712251 s/step
Meep progress: 34.083333333333336/50.0 = 68.2% done in 149.3s, 69.7s to go
on time step 2045 (time=34.0833), 0.0706418 s/step
ex:, 35.0, 0.13850695122294548
Meep progress: 35.03333333333333/50.0 = 70.1% done in 153.3s, 65.5s to go
on time step 2102 (time=35.0333), 0.0711154 s/step
Meep progress: 35.983333333333334/50.0 = 72.0% done in 157.4s, 61.3s to go
on time step 2159 (time=35.9833), 0.0713521 s/step
Meep progress: 36.93333333333333/50.0 = 73.9% done in 161.4s, 57.1s to go
on time step 2216 (time=36.9333), 0.0705946 s/step
Meep progress: 37.88333333333333/50.0 = 75.8% done in 165.5s, 52.9s to go
on time step 2273 (time=37.8833), 0.0707061 s/step
Meep progress: 38.81666666666666/50.0 = 77.6% done in 169.5s, 48.8s to go
on time step 2329 (time=38.8167), 0.0716441 s/step
Meep progress: 39.75/50.0 = 79.5% done in 173.5s, 44.7s to go
on time step 2385 (time=39.75), 0.0719617 s/step
ex:, 40.0, 0.14552885528463194
Meep progress: 40.7/50.0 = 81.4% done in 177.6s, 40.6s to go
on time step 2442 (time=40.7), 0.0713272 s/step
Meep progress: 41.65/50.0 = 83.3% done in 181.6s, 36.4s to go
on time step 2499 (time=41.65), 0.0707835 s/step
Meep progress: 42.6/50.0 = 85.2% done in 185.7s, 32.3s to go
on time step 2556 (time=42.6), 0.0712435 s/step
Meep progress: 43.55/50.0 = 87.1% done in 189.7s, 28.1s to go
on time step 2613 (time=43.55), 0.0711772 s/step
Meep progress: 44.5/50.0 = 89.0% done in 193.8s, 23.9s to go
on time step 2670 (time=44.5), 0.0707421 s/step
ex:, 45.0, 0.11463516422951753
Meep progress: 45.45/50.0 = 90.9% done in 197.8s, 19.8s to go
on time step 2727 (time=45.45), 0.0710496 s/step
Meep progress: 46.4/50.0 = 92.8% done in 201.8s, 15.7s to go
on time step 2784 (time=46.4), 0.0706162 s/step
Meep progress: 47.35/50.0 = 94.7% done in 205.9s, 11.5s to go
on time step 2841 (time=47.35), 0.0707946 s/step
Meep progress: 48.3/50.0 = 96.6% done in 209.9s, 7.4s to go
on time step 2898 (time=48.3), 0.0711866 s/step
Meep progress: 49.25/50.0 = 98.5% done in 214.0s, 3.3s to go
on time step 2955 (time=49.25), 0.0711244 s/step
ex:, 50.0, 0.1078903534460964
run 0 finished at t = 50.0 (3000 timesteps)

Elapsed run time = 222.7804 s
```

**Geometry (no symmetry)**
![hexagon_prism_nosym](https://user-images.githubusercontent.com/7152530/43424948-e4592654-9405-11e8-9305-33044ecaeeb7.png)

**Output (symmetry in `X` and `Y`)**
```
Using MPI version 3.0, 1 processes
-----------
Initializing structure...
Halving computational cell along direction x
Halving computational cell along direction y
Working in 3D dimensions.
Computational cell is 5 x 5 x 5 with resolution 30
     prism, center = (0,0,0.5)
          height 1, axis (0,0,1), 6 vertices:
          {-1.000000e+00,5.551115e-17,0.000000e+00}
          {-5.000000e-01,8.660254e-01,0.000000e+00}
          {5.000000e-01,8.660254e-01,0.000000e+00}
          {1.000000e+00,-5.551115e-17,0.000000e+00}
          {5.000000e-01,-8.660254e-01,0.000000e+00}
          {-5.000000e-01,-8.660254e-01,0.000000e+00}
          dielectric constant epsilon diagonal = (12.25,12.25,12.25)
time for set_epsilon = 1.50421 s
-----------
Meep progress: 2.25/50.0 = 4.5% done in 4.0s, 85.2s to go
on time step 152 (time=2.53333), 0.0264119 s/step
ex:, 5.0, -2.7087637999049014e-05
Meep progress: 5.766666666666667/50.0 = 11.5% done in 8.0s, 61.6s to go
on time step 363 (time=6.05), 0.0190078 s/step
Meep progress: 9.3/50.0 = 18.6% done in 12.0s, 52.7s to go
on time step 574 (time=9.56667), 0.0189615 s/step
ex:, 10.0, -0.001126317771419412
Meep progress: 12.733333333333333/50.0 = 25.5% done in 16.1s, 47.0s to go
on time step 780 (time=13), 0.0194873 s/step
ex:, 15.0, -0.017213201058505788
Meep progress: 16.216666666666665/50.0 = 32.4% done in 20.1s, 41.8s to go
on time step 989 (time=16.4833), 0.0191808 s/step
Meep progress: 19.65/50.0 = 39.3% done in 24.1s, 37.2s to go
on time step 1195 (time=19.9167), 0.0195019 s/step
ex:, 20.0, -0.09466686777357108
Meep progress: 23.03333333333333/50.0 = 46.1% done in 28.1s, 32.9s to go
on time step 1398 (time=23.3), 0.019777 s/step
ex:, 25.0, -0.170966255920729
Meep progress: 26.55/50.0 = 53.1% done in 32.1s, 28.4s to go
on time step 1610 (time=26.8333), 0.0189217 s/step
Meep progress: 29.966666666666665/50.0 = 59.9% done in 36.1s, 24.1s to go
ex:, 30.0, -0.036340185573469866
on time step 1814 (time=30.2333), 0.019645 s/step
Meep progress: 33.46666666666667/50.0 = 66.9% done in 40.1s, 19.8s to go
on time step 2025 (time=33.75), 0.0190118 s/step
ex:, 35.0, 0.13881220324772908
Meep progress: 37.0/50.0 = 74.0% done in 44.1s, 15.5s to go
on time step 2237 (time=37.2833), 0.0188947 s/step
ex:, 40.0, 0.14517834283715125
Meep progress: 40.53333333333333/50.0 = 81.1% done in 48.1s, 11.2s to go
on time step 2449 (time=40.8167), 0.0189472 s/step
Meep progress: 44.05/50.0 = 88.1% done in 52.1s, 7.0s to go
on time step 2660 (time=44.3333), 0.0189864 s/step
ex:, 45.0, 0.11369663857957729
Meep progress: 47.6/50.0 = 95.2% done in 56.2s, 2.8s to go
on time step 2872 (time=47.8667), 0.0188768 s/step
ex:, 50.0, 0.10763175167097991
run 0 finished at t = 50.0 (3000 timesteps)

Elapsed run time = 60.4060 s
```

**Geometry (symmetry in `X` and `Y`)**
![hexagon_prism_sym_xy](https://user-images.githubusercontent.com/7152530/43424961-f7d22898-9405-11e8-90f6-315d0af5a09a.png)

**Output (symmetry in `X`, `Y`, and `Z`)**
```
Using MPI version 3.0, 1 processes
-----------
Initializing structure...
Halving computational cell along direction x
Halving computational cell along direction y
Halving computational cell along direction z
Working in 3D dimensions.
Computational cell is 5 x 5 x 5 with resolution 30
     prism, center = (0,0,0.5)
          height 1, axis (0,0,1), 6 vertices:
          {-1.000000e+00,5.551115e-17,0.000000e+00}
          {-5.000000e-01,8.660254e-01,0.000000e+00}
          {5.000000e-01,8.660254e-01,0.000000e+00}
          {1.000000e+00,-5.551115e-17,0.000000e+00}
          {5.000000e-01,-8.660254e-01,0.000000e+00}
          {-5.000000e-01,-8.660254e-01,0.000000e+00}
          dielectric constant epsilon diagonal = (12.25,12.25,12.25)
time for set_epsilon = 0.655153 s
-----------
ex:, 5.0, 6.344635461642745e-05
Meep progress: 5.8/50.0 = 11.6% done in 4.0s, 30.5s to go
on time step 348 (time=5.8), 0.011498 s/step
ex:, 10.0, 0.0022185346557983205
Meep progress: 12.45/50.0 = 24.9% done in 8.0s, 24.2s to go
on time step 747 (time=12.45), 0.0100467 s/step
ex:, 15.0, 0.028472772371336347
Meep progress: 19.333333333333332/50.0 = 38.7% done in 12.0s, 19.1s to go
on time step 1160 (time=19.3333), 0.00969812 s/step
ex:, 20.0, 0.13390190756857862
ex:, 25.0, 0.23053188817696185
Meep progress: 26.28333333333333/50.0 = 52.6% done in 16.0s, 14.5s to go
on time step 1577 (time=26.2833), 0.00959297 s/step
ex:, 30.0, 0.145030080104607
Meep progress: 33.25/50.0 = 66.5% done in 20.0s, 10.1s to go
on time step 1995 (time=33.25), 0.00957086 s/step
ex:, 35.0, 0.033171166842913505
ex:, 40.0, 0.00276639269136829
Meep progress: 40.2/50.0 = 80.4% done in 24.0s, 5.9s to go
on time step 2412 (time=40.2), 0.0096145 s/step
ex:, 45.0, 0.0002015702868028624
Meep progress: 47.15/50.0 = 94.3% done in 28.0s, 1.7s to go
on time step 2829 (time=47.15), 0.00959507 s/step
ex:, 50.0, 0.00011521947094192625
run 0 finished at t = 50.0 (3000 timesteps)

Elapsed run time = 30.3405 s
```

**Geometry (symmetry in `X`, `Y`, and `Z`)**
Note that the height of the prism is different from the other two examples.
![hexagon_prism_sym_xyz](https://user-images.githubusercontent.com/7152530/43425001-17709234-9406-11e8-971f-a43e6d640d70.png)

**Comment by oskooi** [comment]
Even after shifting the prism object such that its centroid is at (0,0,0), similar to the blocks representation, there is a difference in the epsilon output for the two representations as shown below.

```py
import meep as mp
import math

cell_size = mp.Vector3(5,5,5)

Si = mp.Medium(index=3.5)

use_prism = False

if use_prism:
    z = -0.5
    vertices = [mp.Vector3(-1,0,z),
                mp.Vector3(-0.5,+math.sqrt(3)/2,z),
		mp.Vector3(+0.5,+math.sqrt(3)/2,z),
                mp.Vector3(+1,0,z),
                mp.Vector3(+0.5,-math.sqrt(3)/2,z),
                mp.Vector3(-0.5,-math.sqrt(3)/2,z)]

    geometry = [mp.Prism(vertices, height=1.0, material=Si)]
else:
    geometry = [mp.Block(material=Si, size=mp.Vector3(1,math.sqrt(3),1), center=mp.Vector3()),
                mp.Block(material=Si, size=mp.Vector3(1,math.sqrt(3),1), center=mp.Vector3(),
                         e1=mp.Vector3(0.5,+math.sqrt(3)/2,0), e2=mp.Vector3(-math.sqrt(3)/2,0.5,0), e3=mp.Vector3(0,0,1)),
		mp.Block(material=Si, size=mp.Vector3(1,math.sqrt(3),1), center=mp.Vector3(),
                         e1=mp.Vector3(0.5,-math.sqrt(3)/2,0), e2=mp.Vector3(+math.sqrt(3)/2,0.5,0), e3=mp.Vector3(0,0,1))]

boundary_layers = [mp.PML(thickness=1.0)]

fcen = 1.0
sources = [mp.Source(mp.GaussianSource(fcen, fwidth=0.2*fcen), component=mp.Ez, center=mp.Vector3())]

sim = mp.Simulation(resolution=50,
                    eps_averaging=False,
                    geometry=geometry,
                    cell_size=cell_size,
                    boundary_layers=boundary_layers,
                    sources=sources)

sim.run(mp.at_beginning(mp.output_epsilon), until=0)
```

![hexagon_epsilon_diff](https://user-images.githubusercontent.com/7152530/43483740-4e7421d6-94c1-11e8-8dc1-3ee54e8d2ebc.png)

**Comment by oskooi** [comment]
Turns out that the hexagonal prism was not centered on the origin. Fixed by #441.
